numpy/core/src/multiarray/iterators.c

```diff
@@ -627,9 +662,22 @@ iter_subscript(PyArrayIterObject *self, PyObject *ind)
             goto fail;
         }
         dptr = PyArray_DATA(ret);
-        copyswap = PyArray_DESCR(ret)->f->copyswap;
+        NPY_ARRAYMETHOD_FLAGS transfer_flags = 0;
+        npy_intp one = 1;
+        npy_intp transfer_strides[2] = {size, size};
+        /* We can assume the newly allocated output array is aligned */
+        int is_aligned = IsUintAligned(self->ao);
+        if (PyArray_GetDTypeTransferFunction(
+                    is_aligned, size, size, dtype, dtype, 0, &cast_info,
+                    &transfer_flags) < 0) {
+            goto fail;
+        }
         while (n_steps--) {
-            copyswap(dptr, self->dataptr, 0, ret);
+            char *args[2] = {self->dataptr, dptr};
+            if (cast_info.func(&cast_info.context, args, &one,
+                               transfer_strides, cast_info.auxdata) < 0) {
+                goto fail;
+            }
             start += step_size;
             PyArray_ITER_GOTO1D(self, start);
             dptr += size;
```

Comment:
Cleanup seems also broken in this path, maybe a `goto finish`?  (return 3 lines down)